internal/proxy/multi_rate_limiter.go

```diff
@@ -49,80 +50,98 @@ func GetQuotaErrorString(errCode commonpb.ErrorCode) string {
+func (m *MultiRateLimiter) Check(collectionID int64, rt internalpb.RateType, n int) commonpb.ErrorCode {
 	if !Params.QuotaConfig.QuotaAndLimitsEnabled.GetAsBool() {
 		return commonpb.ErrorCode_Success
 	}
-	limit, rate := m.globalRateLimiter.limit(rt, n)
+
+	limiter := m.collectionLimiters[collectionID]
+	if limiter == nil {
+		return commonpb.ErrorCode_Success
+	}
+
+	limit, rate := limiter.limit(rt, n)
 	if rate == 0 {
-		return m.GetErrorCode(rt)
+		return limiter.getErrorCode(rt)
 	}
 	if limit {
 		return commonpb.ErrorCode_RateLimit
 	}
 	return commonpb.ErrorCode_Success
 }
 
-func (m *MultiRateLimiter) GetErrorCode(rt internalpb.RateType) commonpb.ErrorCode {
-	switch rt {
-	case internalpb.RateType_DMLInsert, internalpb.RateType_DMLDelete, internalpb.RateType_DMLBulkLoad:
-		m.quotaStatesMu.RLock()
-		defer m.quotaStatesMu.RUnlock()
-		return m.quotaStates[milvuspb.QuotaState_DenyToWrite]
-	case internalpb.RateType_DQLSearch, internalpb.RateType_DQLQuery:
-		m.quotaStatesMu.RLock()
-		defer m.quotaStatesMu.RUnlock()
-		return m.quotaStates[milvuspb.QuotaState_DenyToRead]
-	}
-	return commonpb.ErrorCode_Success
-}
-
 // GetQuotaStates returns quota states.
 func (m *MultiRateLimiter) GetQuotaStates() ([]milvuspb.QuotaState, []string) {
 	m.quotaStatesMu.RLock()
 	defer m.quotaStatesMu.RUnlock()
-	states := make([]milvuspb.QuotaState, 0, len(m.quotaStates))
-	reasons := make([]string, 0, len(m.quotaStates))
-	for k, v := range m.quotaStates {
-		states = append(states, k)
-		reasons = append(reasons, GetQuotaErrorString(v))
+	serviceStates := make(map[milvuspb.QuotaState]typeutil.Set[commonpb.ErrorCode])
+
+	// deduplicate same (state, code) pair from different collection
+	for _, limiter := range m.collectionLimiters {
+		limiter.quotaStates.Range(func(key, value interface{}) bool {
+			state := key.(milvuspb.QuotaState)
+			errCode := value.(commonpb.ErrorCode)
+			if serviceStates[state] == nil {
+				serviceStates[state] = typeutil.NewSet[commonpb.ErrorCode]()
+			}
+			serviceStates[state].Insert(errCode)
+			return true
+		})
+	}
+
+	states := make([]milvuspb.QuotaState, 0)
+	reasons := make([]string, 0)
+	for state, errCodes := range serviceStates {
+		for errCode := range errCodes {
+			states = append(states, state)
+			reasons = append(reasons, GetQuotaErrorString(errCode))
+		}
 	}
+
 	return states, reasons
 }
 
 // SetQuotaStates sets quota states for MultiRateLimiter.
-func (m *MultiRateLimiter) SetQuotaStates(states []milvuspb.QuotaState, codes []commonpb.ErrorCode) {
+func (m *MultiRateLimiter) SetRates(rates []*proxypb.CollectionRate) error {
 	m.quotaStatesMu.Lock()
 	defer m.quotaStatesMu.Unlock()
-	m.quotaStates = make(map[milvuspb.QuotaState]commonpb.ErrorCode, len(states))
-	for i := 0; i < len(states); i++ {
-		m.quotaStates[states[i]] = codes[i]
+	m.collectionLimiters = make(map[int64]*rateLimiter, 0)
+	for _, collectionRates := range rates {
+		rateLimiter := newRateLimiter()
```

Comment:
Creating a new limiter for every setting rate may have too much overhead. Maybe we can create a limiter only if it does not already exist, and remove it if it is not in the current collection of rates.